File: baml_language/crates/bridge_go/pkg/tests/call_function_test.go
package tests

import (
	"context"
	"testing"

	"bridge_go/pkg"
)

const callFunctionBamlSource = `
function ReturnOne() -> int {
  1
}

function Identity(s: string) -> string {
  s
}

enum Color {
  Red
  Blue
}

function PickColor(n: int) -> Color {
  if (n == 0) { Color.Red } else { Color.Blue }
}

class Point {
  x int
  y int
}

function MakePoint(x: int, y: int) -> Point {
  Point { x: x, y: y }
}

class Address {
  city string
  zip int
}

class Contact {
  name string
  address Address
}

function MakeContact(name: string, city: string, zip: int) -> Contact {
  Contact { name: name, address: Address { city: city, zip: zip } }
}

function IntOrString(useInt: bool, n: int) -> int | string {
  if (useInt) { n } else { "hello" }
}

function MakeIntList(a: int, b: int, c: int) -> int[] {
  [a, b, c]
}

function MakeStringMap() -> map<string, int> {
  { "a": 1, "b": 2, "c": 3 }
}
`

var testRT *pkg.BamlRuntime

func getTestRuntime(t *testing.T) *pkg.BamlRuntime {
	t.Helper()
	if testRT != nil {
		return testRT
	}
	rt, err := pkg.NewRuntime(".", map[string]string{"main.baml": callFunctionBamlSource})
	if err != nil {
		t.Fatalf("NewRuntime failed: %v", err)
	}
	testRT = rt
	return rt
}

func TestReturnInt(t *testing.T) {
	rt := getTestRuntime(t)
	result, err := rt.CallFunction(context.Background(), "ReturnOne", nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result != int64(1) {
		t.Fatalf("expected 1, got %v (%T)", result, result)
	}
}

func TestIdentityString(t *testing.T) {
	rt := getTestRuntime(t)
	result, err := rt.CallFunction(context.Background(), "Identity", map[string]any{"s": "hello"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result != "hello" {
		t.Fatalf("expected 'hello', got %v", result)
	}
}

func TestEnumResult(t *testing.T) {
	rt := getTestRuntime(t)
	result, err := rt.CallFunction(context.Background(), "PickColor", map[string]any{"n": 0})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	e, ok := result.(pkg.DynamicEnum)
	if !ok {
		t.Fatalf("expected DynamicEnum, got %T: %v", result, result)
	}
	if e.Name != "Color" {
		t.Fatalf("expected enum name 'Color', got %q", e.Name)
	}
	if e.Value != "Red" {
		t.Fatalf("expected enum value 'Red', got %q", e.Value)
	}
}

func TestClassResult(t *testing.T) {
	rt := getTestRuntime(t)
	result, err := rt.CallFunction(context.Background(), "MakePoint", map[string]any{"x": 10, "y": 20})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	c, ok := result.(pkg.DynamicClass)
	if !ok {
		t.Fatalf("expected DynamicClass, got %T: %v", result, result)
	}
	if c.Name != "Point" {
		t.Fatalf("expected class name 'Point', got %q", c.Name)
	}
	if c.Fields["x"] != int64(10) || c.Fields["y"] != int64(20) {
		t.Fatalf("expected {x:10, y:20}, got %v", c.Fields)
	}
}

func TestNestedClass(t *testing.T) {
	rt := getTestRuntime(t)
	result, err := rt.CallFunction(context.Background(), "MakeContact", map[string]any{
		"name": "Bob", "city": "Seattle", "zip": 98101,
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	c, ok := result.(pkg.DynamicClass)
	if !ok {
		t.Fatalf("expected DynamicClass, got %T: %v", result, result)
	}
	if c.Name != "Contact" {
		t.Fatalf("expected class name 'Contact', got %q", c.Name)
	}
	if c.Fields["name"] != "Bob" {
		t.Fatalf("expected name=Bob, got %v", c.Fields["name"])
	}
	addr, ok := c.Fields["address"].(pkg.DynamicClass)
	if !ok {
		t.Fatalf("expected address to be DynamicClass, got %T", c.Fields["address"])
	}
	if addr.Fields["city"] != "Seattle" || addr.Fields["zip"] != int64(98101) {
		t.Fatalf("expected address={city:Seattle, zip:98101}, got %v", addr.Fields)
	}
}

func TestUnionIntOrString(t *testing.T) {
	rt := getTestRuntime(t)

	// Int branch
	result, err := rt.CallFunction(context.Background(), "IntOrString", map[string]any{"useInt": true, "n": 42})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	u, ok := result.(pkg.DynamicUnion)
	if !ok {
		t.Fatalf("expected DynamicUnion, got %T: %v", result, result)
	}
	if u.Value != int64(42) {
		t.Fatalf("expected 42, got %v (%T)", u.Value, u.Value)
	}

	// String branch
	result, err = rt.CallFunction(context.Background(), "IntOrString", map[string]any{"useInt": false, "n": 0})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	u, ok = result.(pkg.DynamicUnion)
	if !ok {
		t.Fatalf("expected DynamicUnion, got %T: %v", result, result)
	}
	if u.Value != "hello" {
		t.Fatalf("expected 'hello', got %v", u.Value)
	}
}

func TestListReturn(t *testing.T) {
	rt := getTestRuntime(t)
	result, err := rt.CallFunction(context.Background(), "MakeIntList", map[string]any{"a": 1, "b": 2, "c": 3})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	list, ok := result.([]any)
	if !ok {
		t.Fatalf("expected []any, got %T", result)
	}
	if len(list) != 3 || list[0] != int64(1) || list[1] != int64(2) || list[2] != int64(3) {
		t.Fatalf("expected [1,2,3], got %v", list)
	}
}

func TestMapReturn(t *testing.T) {
	rt := getTestRuntime(t)
	result, err := rt.CallFunction(context.Background(), "MakeStringMap", nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	m, ok := result.(map[string]any)
	if !ok {
		t.Fatalf("expected map, got %T", result)
	}
	if m["a"] != int64(1) || m["b"] != int64(2) || m["c"] != int64(3) {
		t.Fatalf("expected {a:1, b:2, c:3}, got %v", m)
	}
}

func TestUnknownFunction(t *testing.T) {
	rt := getTestRuntime(t)
	_, err := rt.CallFunction(context.Background(), "NonExistentFunction", nil)
	if err == nil {
		t.Fatal("expected error for unknown function")
	}
}

File: baml_language/sdks/typescript/bridge_typescript/tests/call_function.test.ts
/**
 * Tests for bridge_typescript: callFunctionSync and callFunction (async) covering
 * classes, enums, unions, lists, maps, nested classes, and error cases.
 *
 * All BAML source is embedded inline. No LLM calls — only pure expression
 * functions — so these run without API keys.
 */

import { BamlRuntime } from '../dist/native.js';
import { callFunctionSync, callFunction, FunctionResult } from '../dist/index.js';

// ============================================================================
// BAML source used by tests
// ============================================================================

const BAML_SOURCE = `
// ── Simple scalars (keep two for smoke testing) ──

function ReturnOne() -> int {
    1
}

function Identity(s: string) -> string {
    s
}

// ── Enums ──

enum Color {
    Red,
    Green,
    Blue,
}

function PickColor(x: int) -> Color {
    if (x > 0) { Color.Red } else { Color.Blue }
}

enum Status {
    Active,
    Inactive,
    Pending,
}

function GetStatus(active: bool) -> Status {
    if (active) { Status.Active } else { Status.Inactive }
}

// ── Classes ──

class Point {
    x int
    y int
}

function MakePoint(x: int, y: int) -> Point {
    Point { x: x, y: y }
}

class Person {
    name string
    age int
    active bool
}

function MakePerson(n: string, a: int) -> Person {
    Person { name: n, age: a, active: true }
}

// ── Nested classes ──

class Address {
    city string
    zip int
}

class Contact {
    name string
    address Address
}

function MakeContact(name: string, city: string, zip: int) -> Contact {
    Contact { name: name, address: Address { city: city, zip: zip } }
}

// ── Unions ──

function IntOrString(asInt: bool, x: int) -> int | string {
    if (asInt) { x } else { "hello" }
}

function ColorOrPoint(useColor: bool) -> Color | Point {
    if (useColor) { Color.Green } else { Point { x: 1, y: 2 } }
}

function ClassifyAmbiguousEmptyList(value: int[] | string[]) -> string {
    match (value) {
        let ints: int[] => "ints",
        let strings: string[] => "strings",
    }
}

// ── Lists and maps ──

function MakeIntList(a: int, b: int, c: int) -> int[] {
    [a, b, c]
}

function MakeStringMap() -> map<string, int> {
    {"a": 1, "b": 2, "c": 3}
}

// ── Class with list field ──

class Tagged {
    label string
    scores int[]
}

function MakeTagged(label: string) -> Tagged {
    Tagged { label: label, scores: [10, 20, 30] }
}

// ── Bigint round-trip ──

function EchoBigint(x: bigint) -> bigint {
    x
}
`;

// ============================================================================
// Helper
// ============================================================================

function makeRuntime(bamlSource: string): BamlRuntime {
    return BamlRuntime.initializeRuntime('.', { 'main.baml': bamlSource });
}

// ============================================================================
// Tests
// ============================================================================

describe('callFunctionSync', () => {
    let rt: BamlRuntime;

    beforeAll(() => {
        rt = makeRuntime(BAML_SOURCE);
    });

    // ── Smoke tests (simple scalars) ──

    it('ReturnOne → 1', () => {
        const result = callFunctionSync(rt, 'ReturnOne', {});
        expect(result).toBeInstanceOf(FunctionResult);
        expect(result.result()).toBe(1);
    });

    it('Identity("hello") → "hello"', () => {
        const result = callFunctionSync(rt, 'Identity', { s: 'hello' });
        expect(result.result()).toBe('hello');
    });

    // ── Enum results ──

    it('PickColor(1) → "Red"', () => {
        const result = callFunctionSync(rt, 'PickColor', { x: 1 });
        expect(result.result()).toBe('Red');
    });

    it('PickColor(-1) → "Blue"', () => {
        const result = callFunctionSync(rt, 'PickColor', { x: -1 });
        expect(result.result()).toBe('Blue');
    });

    it('GetStatus(true) → "Active"', () => {
        const result = callFunctionSync(rt, 'GetStatus', { active: true });
        expect(result.result()).toBe('Active');
    });

    it('GetStatus(false) → "Inactive"', () => {
        const result = callFunctionSync(rt, 'GetStatus', { active: false });
        expect(result.result()).toBe('Inactive');
    });

    // ── Class results ──

    it('MakePoint(3, 4) → { x: 3, y: 4 }', () => {
        const result = callFunctionSync(rt, 'MakePoint', { x: 3, y: 4 });
        expect(result.result()).toEqual({ x: 3, y: 4 });
    });

    it('MakePerson("Alice", 30) → { name, age, active }', () => {
        const result = callFunctionSync(rt, 'MakePerson', { n: 'Alice', a: 30 });
        expect(result.result()).toEqual({
            name: 'Alice',
            age: 30,
            active: true,
        });
    });

    // ── Nested class results ──

    it('MakeContact returns nested class', () => {
        const result = callFunctionSync(rt, 'MakeContact', {
            name: 'Bob',
            city: 'Seattle',
            zip: 98101,
        });
        expect(result.result()).toEqual({
            name: 'Bob',
            address: { city: 'Seattle', zip: 98101 },
        });
    });

    // ── Union results ──

    it('IntOrString(true, 42) → int branch', () => {
        const result = callFunctionSync(rt, 'IntOrString', { asInt: true, x: 42 });
        expect(result.result()).toBe(42);
    });

    it('IntOrString(false, 42) → string branch', () => {
        const result = callFunctionSync(rt, 'IntOrString', { asInt: false, x: 42 });
        expect(result.result()).toBe('hello');
    });

    it('ColorOrPoint(true) → enum branch (Color.Green)', () => {
        const result = callFunctionSync(rt, 'ColorOrPoint', { useColor: true });
        expect(result.result()).toBe('Green');
    });

    it('ColorOrPoint(false) → class branch (Point)', () => {
        const result = callFunctionSync(rt, 'ColorOrPoint', { useColor: false });
        expect(result.result()).toEqual({ x: 1, y: 2 });
    });

    it('raw [] uses the dynamic default for int[] | string[]', () => {
        const result = callFunctionSync(rt, 'ClassifyAmbiguousEmptyList', { value: [] });
        expect(result.result()).toBe('ints');
    });

    // ── Lists and maps ──

    it('MakeIntList(1, 2, 3) → [1, 2, 3]', () => {
        const result = callFunctionSync(rt, 'MakeIntList', { a: 1, b: 2, c: 3 });
        expect(result.result()).toEqual([1, 2, 3]);
    });

    it('MakeStringMap() → { a: 1, b: 2, c: 3 }', () => {
        const result = callFunctionSync(rt, 'MakeStringMap', {});
        expect(result.result()).toEqual({ a: 1, b: 2, c: 3 });
    });

    // ── Class with list field ──

    it('MakeTagged returns class with list field', () => {
        const result = callFunctionSync(rt, 'MakeTagged', { label: 'test' });
        expect(result.result()).toEqual({
            label: 'test',
            scores: [10, 20, 30],
        });
    });

    // ── Bigint round-trip ──

    it('EchoBigint round-trips a small bigint', () => {
        const result = callFunctionSync(rt, 'EchoBigint', { x: 42n });
        const value = result.result();
        expect(typeof value).toBe('bigint');
        expect(value).toBe(42n);
    });

    it('EchoBigint round-trips a large positive bigint', () => {
        const huge = 99999999999999999999n;
        const result = callFunctionSync(rt, 'EchoBigint', { x: huge });
        const value = result.result();
        expect(typeof value).toBe('bigint');
        expect(value).toBe(huge);
    });

    it('EchoBigint round-trips a negative bigint', () => {
        const result = callFunctionSync(rt, 'EchoBigint', { x: -42n });
        const value = result.result();
        expect(typeof value).toBe('bigint');
        expect(value).toBe(-42n);
    });

    // ── Error case ──

    it('throws for unknown function', () => {
        expect(() => callFunctionSync(rt, 'NonExistent', {})).toThrow();
    });
});

describe('callFunction (async)', () => {
    let rt: BamlRuntime;

    beforeAll(() => {
        rt = makeRuntime(BAML_SOURCE);
    });

    // ── Smoke tests (simple scalars) ──

    it('ReturnOne → 1', async () => {
        const result = await callFunction(rt, 'ReturnOne', {});
        expect(result).toBeInstanceOf(FunctionResult);
        expect(result.result()).toBe(1);
    });

    it('Identity("hello") → "hello"', async () => {
        const result = await callFunction(rt, 'Identity', { s: 'hello' });
        expect(result.result()).toBe('hello');
    });

    // ── Enum results ──

    it('PickColor(1) → "Red"', async () => {
        const result = await callFunction(rt, 'PickColor', { x: 1 });
        expect(result.result()).toBe('Red');
    });

    it('PickColor(-1) → "Blue"', async () => {
        const result = await callFunction(rt, 'PickColor', { x: -1 });
        expect(result.result()).toBe('Blue');
    });

    it('GetStatus(true) → "Active"', async () => {
        const result = await callFunction(rt, 'GetStatus', { active: true });
        expect(result.result()).toBe('Active');
    });

    it('GetStatus(false) → "Inactive"', async () => {
        const result = await callFunction(rt, 'GetStatus', { active: false });
        expect(result.result()).toBe('Inactive');
    });

    // ── Class results ──

    it('MakePoint(3, 4) → { x: 3, y: 4 }', async () => {
        const result = await callFunction(rt, 'MakePoint', { x: 3, y: 4 });
        expect(result.result()).toEqual({ x: 3, y: 4 });
    });

    it('MakePerson("Alice", 30) → { name, age, active }', async () => {
        const result = await callFunction(rt, 'MakePerson', { n: 'Alice', a: 30 });
        expect(result.result()).toEqual({
            name: 'Alice',
            age: 30,
            active: true,
        });
    });

    // ── Nested class results ──

    it('MakeContact returns nested class', async () => {
        const result = await callFunction(rt, 'MakeContact', {
            name: 'Bob',
            city: 'Seattle',
            zip: 98101,
        });
        expect(result.result()).toEqual({
            name: 'Bob',
            address: { city: 'Seattle', zip: 98101 },
        });
    });

    // ── Union results ──

    it('IntOrString(true, 42) → int branch', async () => {
        const result = await callFunction(rt, 'IntOrString', { asInt: true, x: 42 });
        expect(result.result()).toBe(42);
    });

    it('IntOrString(false, 42) → string branch', async () => {
        const result = await callFunction(rt, 'IntOrString', { asInt: false, x: 42 });
        expect(result.result()).toBe('hello');
    });

    it('ColorOrPoint(true) → enum branch (Color.Green)', async () => {
        const result = await callFunction(rt, 'ColorOrPoint', { useColor: true });
        expect(result.result()).toBe('Green');
    });

    it('ColorOrPoint(false) → class branch (Point)', async () => {
        const result = await callFunction(rt, 'ColorOrPoint', { useColor: false });
        expect(result.result()).toEqual({ x: 1, y: 2 });
    });

    // ── Lists and maps ──

    it('MakeIntList(1, 2, 3) → [1, 2, 3]', async () => {
        const result = await callFunction(rt, 'MakeIntList', { a: 1, b: 2, c: 3 });
        expect(result.result()).toEqual([1, 2, 3]);
    });

    it('MakeStringMap() → { a: 1, b: 2, c: 3 }', async () => {
        const result = await callFunction(rt, 'MakeStringMap', {});
        expect(result.result()).toEqual({ a: 1, b: 2, c: 3 });
    });

    // ── Class with list field ──

    it('MakeTagged returns class with list field', async () => {
        const result = await callFunction(rt, 'MakeTagged', { label: 'test' });
        expect(result.result()).toEqual({
            label: 'test',
            scores: [10, 20, 30],
        });
    });

    // ── Bigint round-trip ──

    it('EchoBigint round-trips a small bigint', async () => {
        const result = await callFunction(rt, 'EchoBigint', { x: 42n });
        const value = result.result();
        expect(typeof value).toBe('bigint');
        expect(value).toBe(42n);
    });

    it('EchoBigint round-trips a large positive bigint', async () => {
        const huge = 99999999999999999999n;
        const result = await callFunction(rt, 'EchoBigint', { x: huge });
        const value = result.result();
        expect(typeof value).toBe('bigint');
        expect(value).toBe(huge);
    });

    it('EchoBigint round-trips a negative bigint', async () => {
        const result = await callFunction(rt, 'EchoBigint', { x: -42n });
        const value = result.result();
        expect(typeof value).toBe('bigint');
        expect(value).toBe(-42n);
    });

    // ── Error case ──

    it('rejects for unknown function', async () => {
        await expect(callFunction(rt, 'NonExistent', {})).rejects.toThrow();
    });
});
